File: Lintcode/1246. Longest Repeating Character Replacement/LintCode_1246.java
public class Solution {
    /**
     * @param s: a string
     * @param k: a integer
     * @return: return a integer
     */
    public int characterReplacement(String s, int k) {
        // write your code here
        if (s == null) {
            return 0;
        }
        int j = 0, answer = 0, maxFreq = 0, count;
        HashMap<Character, Integer> counter = new HashMap<>();
        
        for (int i = 0; i < s.length(); i++) {
            while (j < s.length() && j - i - maxFreq <= k) {
                count = counter.getOrDefault(s.charAt(j), 0) + 1;
                counter.put(s.charAt(j), count);
                maxFreq = Math.max(maxFreq, count);
                j++;
            }
            if (j - i - maxFreq > k) {
                answer = Math.max(answer, j - i - 1);
            } else {
                answer = Math.max(answer, j - i);
            }
            
            count = counter.get(s.charAt(i)) - 1;
            counter.put(s.charAt(i), count);
            
            maxFreq = getMaxFreq(counter);
        }
        return answer;
    }
}

File: Lintcode/1246. Longest Repeating Character Replacement/LintCode_1246.py
class Solution:
    """
    @param s: a string
    @param k: a integer
    @return: return a integer
    """
    def characterReplacement(self, s, k):
        # write your code here
        counter = {}
        answer = 0
        j = 0
        max_freq = 0
        for i in range(len(s)):
            while j < len(s) and j - i - max_freq <= k:
                counter[s[j]] = counter.get(s[j], 0) + 1
                max_freq = max(max_freq, counter[s[j]])
                j += 1
            if j - i - max_freq > k:
                answer = max(answer, j - i - 1)
            else:
                answer = max(answer, j - i)
            
            counter[s[i]] -= 1
            max_freq = max(counter.values())
        return answer